tools/make_patch.c

```diff
@@ -0,0 +1,476 @@
+#define PROGRAM_NAME "make_patch"
+#define USAGE_OPTS "labels.sym constants.out patched.gbc original.gbc vc.patch.template vc.patch"
+
+#include "common.h"
+
+#include <ctype.h>
+
+struct Buffer {
+	size_t item_size;
+	size_t size;
+	size_t capacity;
+	char *data;
+};
+
+struct Patch {
+	unsigned int offset;
+	unsigned int size;
+};
+
+struct Symbol {
+	struct Symbol *next;
+	unsigned int address;
+	unsigned int offset;
+	char name[]; // VLA
+};
+
+struct Buffer *create_buffer(size_t item_size) {
+	struct Buffer *buffer = xmalloc(sizeof(*buffer));
+	buffer->item_size = item_size;
+	buffer->size = 0;
+	buffer->capacity = 0x10;
+	buffer->data = xmalloc(buffer->capacity * item_size);
+	return buffer;
+}
+
+void append_to_buffer(struct Buffer *buffer, const void *item) {
+	if (buffer->size >= buffer->capacity) {
+		buffer->capacity = (buffer->capacity + 1) * 2;
+		buffer->data = xrealloc(buffer->data, buffer->capacity * buffer->item_size);
+	}
+	memcpy(buffer->data + (buffer->size++ * buffer->item_size), item, buffer->item_size);
+}
+
+void free_buffer(struct Buffer *buffer) {
+	free(buffer->data);
+	free(buffer);
+}
+
+int get_address_type_limit(int address) {
+	return address >= 0x8000 && address < 0xa000 ? 0xa000 // VRAM
+	     : address >= 0xa000 && address < 0xc000 ? 0xc000 // SRAM
+	     : address >= 0xc000 && address < 0xf000 ? 0xd000 // WRAM
+	     : address >= 0xff80 && address < 0xffff ? 0x10000 // HRAM
+	     : 0x4000; // ROM
+}
+
+void append_symbol(struct Symbol **symbols, const char *name, int bank, int address) {
+	struct Symbol *symbol = xmalloc(sizeof(*symbol) + strlen(name) + 1);
+	symbol->address = address;
+	symbol->offset = bank > 0 ? address + (bank - 1) * get_address_type_limit(address) : address;
+	strcpy(symbol->name, name);
+	symbol->next = *symbols;
+	*symbols = symbol;
+}
+
+void free_symbols(struct Symbol *list) {
+	for (struct Symbol *next; list; list = next) {
+		next = list->next;
+		free(list);
+	}
+}
+
+const struct Symbol *find_symbol(const struct Symbol *symbols, const char *name) {
+	size_t namelen = strlen(name);
+	for (const struct Symbol *symbol = symbols; symbol; symbol = symbol->next) {
+		size_t symbolnamelen = strlen(symbol->name);
+		if (namelen > symbolnamelen) {
+			continue;
+		}
+		const char *symbolname = symbol->name;
+		if (name[0] == '.') {
+			// If `name` is a local label, compare it to the local part of `symbol->name`
+			symbolname += symbolnamelen - namelen;
```

Comment:
Is ignoring the global part (and thus matching `My.house` to `[Your].house`) intentional?

Reply:
Yes. The .sym file will always have fully-qualified names, and we only call `find_symbol` to find (a) labels used in templates like `wSomeVar` (which should be written fully-qualified) and (b) the macro-generated `.VC_Foo` and `.VC_Foo_End` labels (which we can't know what their global part is).

Reply:
There probably should be a comment to that effect, then.